### PR TITLE
Refactor: Typescript, promises, models

### PR DESCRIPTION
**Pull request by pekko**
Sorry, it is large.

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -29,10 +31,15 @@ app.use(morgan('combined', {stream: {
   write: function(message) { log.info('HTTP: ' + message); }
 }}));
 
+var dirname = '/Users/plipsanen/personal/nappikauppa2'; // TODO
```
Comment (Blizzara) [inline app.ts]: this ain't cool

File: frontend/src/js/components/Store.tsx [hunk 2]
```diff
@@ -210,9 +252,9 @@ var Store = React.createClass({
           window.location.href = res.url;
```
Comment (Blizzara) [inline frontend/src/js/components/Store.tsx]: the replaceState(this.initialState()); doesn't work anymore

File: frontend/src/js/components/Contacts.tsx [hunk 3]
```diff
@@ -65,42 +80,39 @@ var Contacts = React.createClass({
...
         <h2>Yhteystiedot <small>4/5</small></h2>
         <div>
-          <Input
+          <Bootstrap.Input
             label='Nimi'
             type='text'
             bsStyle={ _.contains(this.state.errors, 'name') ? ('error') : null }
             readOnly={!active}
-            onChange={this.onValueChange.bind(null, 'name')}
+            onChange={this.onValueChange.bind(this, 'name')}
```
Comment (Blizzara) [inline frontend/src/js/components/Contacts.tsx]: was there some reason to replace null with this?

File: frontend/src/js/components/FinalConfirmation.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+export default class FinalConfirmation extends React.Component<IFinalConfirmationProps, any> {
+
+  render() {
+    if(!this.props.order) {
+      return (<div />);
+    }
+    var numTickets = this.props.order.tickets.length;
+    var finalPrice = this.props.order.order_price;
```
Comment (Blizzara) [inline frontend/src/js/components/FinalConfirmation.tsx]: even though order_price seems to come correctly from backend, frontend loses it somehow (checking React console for FinalConfirmation, it shows that this.props.order.order_price = null)

Comment (pekko) [inline frontend/src/js/components/FinalConfirmation.tsx]: yeap, realized that while testing, fix is coming in another branch (sorry this is a bit messed up now, but not for long)

File: app.ts (hunk 1 above)
Comment (pekko) [inline app.ts]: damn

Comment (pekko) [inline app.ts]: hmm `__dirname` seems to work now, I'm pretty sure it was broken in typescript just a moment ago...

**Pull request merged**